third_party/terraform/resources/resource_google_project_services.go

```diff
@@ -48,330 +54,321 @@ func resourceGoogleProjectServices() *schema.Resource {
-
-	// Get services from API
-	apiServices, err := getApiServices(d.Id(), config, ignoreProjectServices)
+	project, err := getProject(d, config)
 	if err != nil {
-		return fmt.Errorf("Error updating services: %v", err)
+		return err
 	}
 
-	// This call disables any APIs that aren't defined in cfgServices,
-	// and enables all of those that are
-	err = reconcileServices(cfgServices, apiServices, config, d.Id())
+	// Get services from config
+	services, err := expandServiceUsageProjectServicesServices(d.Get("services"), d, config)
 	if err != nil {
-		return fmt.Errorf("Error updating services: %v", err)
+		return err
+	}
+
+	log.Printf("[DEBUG]: Updating Project Services %s", d.Id())
+	if err := setServiceUsageProjectEnabledServices(services, project, config); err != nil {
+		return fmt.Errorf("Error disabling services not in config for Project Services %s: %v", project, err)
 	}
+	log.Printf("[DEBUG]: Finished updating Project Services %s", d.Id())
 
 	return resourceGoogleProjectServicesRead(d, meta)
 }
 
 func resourceGoogleProjectServicesDelete(d *schema.ResourceData, meta interface{}) error {
-	log.Printf("[DEBUG]: Deleting google_project_services")
-
 	if disable := d.Get("disable_on_destroy"); !(disable.(bool)) {
-		log.Printf("Not disabling service '%s', because disable_on_destroy is false.", d.Id())
+		log.Printf("[WARN] Project Services disable_on_destroy set to false, skip disabling services for %s.", d.Id())
 		d.SetId("")
 		return nil
 	}
 
 	config := meta.(*Config)
-	services := resourceServices(d)
+
+	// Get services from config
+	services, err := expandServiceUsageProjectServicesServices(d.Get("services"), d, config)
+	if err != nil {
+		return err
+	}
+	project := d.Id()
+
+	log.Printf("[DEBUG]: Disabling Project Services %s: %+v", project, services)
 	for _, s := range services {
-		if err := disableService(s, d.Id(), config, true); err != nil {
-			return err
+		if err := disableServiceUsageProjectService(s, project, config, true); err != nil {
+			return fmt.Errorf("Unable to destroy google_project_services for %s: %s", d.Id(), err)
 		}
 	}
+	log.Printf("[DEBUG] Finished disabling Project Services %s: %+v", project, services)
+
 	d.SetId("")
 	return nil
 }
 
-// This function ensures that the services enabled for a project exactly match that
-// in a config by disabling any services that are returned by the API but not present
-// in the config
-func reconcileServices(cfgServices, apiServices []string, config *Config, pid string) error {
-	// Helper to convert slice to map
-	m := func(vals []string) map[string]struct{} {
-		sm := make(map[string]struct{})
-		for _, s := range vals {
-			sm[s] = struct{}{}
+// setServiceUsageProjectEnabledServices authoritatively sets the enabled
+// services for a set of project services.
+func setServiceUsageProjectEnabledServices(services []string, project string, config *Config) error {
+	enabledSet, err := getEnabledServiceSet(project, config, ignoredProjectServicesSet)
```

Comment:
```suggestion
	currentEnabledServiceSet, err := getEnabledServiceSet(project, config, ignoredProjectServicesSet)
```

Reply:
I (personally) get a little confused seeing `enableServiceUsageProjectServices` followed by `enabledSet`, and then seeing `enabledSet` values getting disabled. Hopefully these or similar names will make it a little more clear what each set is?

I could also just be a bit slow on the uptake, so feel free to ignore both of these 👍

Reply:
i'll rename some things and leave better comments